File: backend/src/svap/stages/stage0_source_fetch.py
"""
Stage 0: Enforcement Source Fetching

Fetches enforcement action documents (DOJ press releases, GAO reports) from
their official government URLs, extracts visible text from the HTML, and
ingests into the RAG store so Stage 1 can extract structured cases.

Sources are read from the enforcement_sources database table. Sources that
already have a document (has_document=True) are skipped. After fetching,
all pending documents are validated by the LLM to confirm they are relevant
enforcement documents, and a summary is generated.

Input:  Enforcement sources from the database (seeded from enforcement_sources.json)
Output: Documents in the RAG store (doc_type='enforcement'), summaries in enforcement_sources
"""

import os
import re
import ssl
import urllib.request
from datetime import UTC, datetime
from html.parser import HTMLParser

from svap.bedrock_client import BedrockClient
from svap.rag import DocumentIngester
from svap.storage import SVAPStorage

CONFIG_BUCKET = os.environ.get("SVAP_CONFIG_BUCKET", "")

# Tags whose content should be skipped entirely
_SKIP_TAGS = frozenset({"script", "style", "noscript", "svg", "head"})

VALIDATION_SYSTEM = (
    "You are an analyst evaluating enforcement documents for a healthcare fraud "
    "analysis pipeline. You determine whether a document describes a real enforcement "
    "action and provide a concise summary."
)


class _HTMLTextExtractor(HTMLParser):
    """Extract visible text from HTML, skipping scripts/styles."""

    def __init__(self):
        super().__init__()
        self._parts: list[str] = []
        self._skip_depth = 0

    def handle_starttag(self, tag, attrs):
        if tag in _SKIP_TAGS:
            self._skip_depth += 1

    def handle_endtag(self, tag):
        if tag in _SKIP_TAGS and self._skip_depth > 0:
            self._skip_depth -= 1

    def handle_data(self, data):
        if self._skip_depth == 0:
            self._parts.append(data)

    def get_text(self) -> str:
        raw = " ".join(self._parts)
        lines = [line.strip() for line in raw.splitlines()]
        text = "\n".join(line for line in lines if line)
        text = re.sub(r"\n{3,}", "\n\n", text)
        return text.strip()


def _fetch_url(url: str, timeout: int = 30) -> str:
    """Fetch a URL and return the response body as text."""
    ctx = ssl.create_default_context()
    req = urllib.request.Request(
        url,
        headers={
            "User-Agent": "SVAP-Pipeline/1.0 (Structural Vulnerability Analysis)",
            "Accept": "text/html,application/xhtml+xml,*/*",
        },
    )
    with urllib.request.urlopen(req, timeout=timeout, context=ctx) as resp:
        charset = resp.headers.get_content_charset() or "utf-8"
        return resp.read().decode(charset, errors="replace")


def _extract_text(html: str) -> str:
    """Extract visible text from HTML content."""
    parser = _HTMLTextExtractor()
    parser.feed(html)
    return parser.get_text()


def _store_to_s3(key: str, body_bytes: bytes, content_type: str):
    """Store a file to the data S3 bucket."""
    if not CONFIG_BUCKET:
        return
    import boto3

    boto3.client("s3").put_object(
        Bucket=CONFIG_BUCKET, Key=key, Body=body_bytes, ContentType=content_type
    )


def _validate_document(client: BedrockClient, source: dict, doc_text: str) -> tuple[str, bool]:
    """Use LLM to validate a document and generate a summary.

    Returns (summary, is_valid).
    """
    text_sample = doc_text[:4000]
    prompt = (
        f"Analyze this enforcement document and provide:\n"
        f"1. A 2-3 sentence summary of the enforcement action described\n"
        f"2. Whether this is a valid healthcare fraud enforcement document (true/false)\n\n"
        f"Document source: {source['name']}\n"
        f"Description: {source.get('description', 'N/A')}\n"
        f"URL: {source.get('url', 'uploaded')}\n\n"
        f"Document text (first portion):\n{text_sample}\n\n"
        f'Respond in JSON: {{"summary": "...", "is_valid": true/false}}'
    )
    result = client.invoke_json(prompt, system=VALIDATION_SYSTEM, temperature=0.1, max_tokens=500)
    return result.get("summary", ""), bool(result.get("is_valid", False))


def _fetch_missing_documents(storage, sources, config):
    """Phase 1: Fetch documents for sources that don't have one yet."""
    ingester = DocumentIngester(storage, config)
    fetched = 0
    skipped = 0
    failed = 0

    for source in sources:
        source_id = source["source_id"]
        name = source["name"]
        url = source.get("url")

        if source["has_document"]:
            print(f"  Skipping (has document): {name}")
            skipped += 1
            continue

        if not url:
            print(f"  Skipping (no URL): {name}")
            skipped += 1
            continue

        print(f"  Fetching: {name}")
        try:
            html = _fetch_url(url)
            text = _extract_text(html)

            if len(text) < 200:
                print(f"    Skipped: text too short ({len(text)} chars)")
                failed += 1
                continue

            doc_id, n_chunks = ingester.ingest_text(
                text=text,
                filename=source_id,
                doc_type="enforcement",
                metadata={
                    "url": url,
                    "source_name": name,
                    "source_id": source_id,
                    "fetch_date": datetime.now(UTC).isoformat(),
                },
            )

            s3_key = f"enforcement-sources/{source_id}/press_release.html"
            _store_to_s3(s3_key, html.encode("utf-8"), "text/html")

            storage.update_enforcement_source_document(source_id, s3_key=s3_key, doc_id=doc_id)
            print(f"    Ingested: {len(text)} chars, {n_chunks} chunks")
            fetched += 1

        except Exception as e:
            print(f"    Failed to fetch {url}: {e}")
            failed += 1

    return fetched, skipped, failed


def _validate_pending_documents(storage, client):
    """Phase 2: Validate all sources that have documents but no summary yet."""
    validated = 0
    sources = storage.get_enforcement_sources()
    docs = storage.get_all_documents(doc_type="enforcement")

    for source in sources:
        if not source["has_document"] or source["validation_status"] != "pending":
            continue

        print(f"  Validating: {source['name']}")
        try:
            doc = next((d for d in docs if d["doc_id"] == source["doc_id"]), None)
            if not doc:
                print("    Document not found in RAG store")
                storage.update_enforcement_source_summary(
                    source["source_id"], "Document not found in RAG store", "error"
                )
                continue

            summary, is_valid = _validate_document(client, source, doc["full_text"])
            status = "valid" if is_valid else "invalid"
            storage.update_enforcement_source_summary(source["source_id"], summary, status)
            print(f"    {status}: {summary[:80]}...")
            validated += 1

        except Exception as e:
            print(f"    Validation failed: {e}")
            storage.update_enforcement_source_summary(
                source["source_id"], f"Validation error: {e}", "error"
            )

    return validated


def run(storage: SVAPStorage, client: BedrockClient, run_id: str, config: dict):
    """
    Execute Stage 0: Fetch enforcement source documents and ingest into RAG store.

    Reads sources from the database. Skips sources that already have documents.
    After fetching, validates all pending documents with the LLM.
    """
    print("Stage 0: Enforcement Source Fetching")
    storage.log_stage_start(run_id, 0)

    try:
        storage.seed_enforcement_sources_if_empty()
        sources = storage.get_enforcement_sources()

        fetched, skipped, failed = _fetch_missing_documents(storage, sources, config)
        validated = _validate_pending_documents(storage, client)

        storage.log_stage_complete(
            run_id,
            0,
            {
                "documents_fetched": fetched,
                "documents_skipped": skipped,
                "documents_failed": failed,
                "documents_validated": validated,
            },
        )
        print(
            f"  Stage 0 complete: {fetched} fetched, {skipped} skipped, "
            f"{failed} failed, {validated} validated."
        )

    except Exception as e:
        storage.log_stage_failed(run_id, 0, str(e))
        raise

File: backend/shared/src/stages/stage0_source_fetch.rs
//! Stage 0: Enforcement Source Fetching
//!
//! Fetches enforcement action documents from URLs, extracts visible text,
//! and ingests into the RAG store. Then validates pending documents via LLM.

use regex::Regex;
use serde_json::json;
use tokio_postgres::Client;
use tracing::{error, info};

use crate::bedrock::BedrockClient;
use crate::db;
use crate::rag::DocumentIngester;
use crate::types::{Config, Document, EnforcementSource};

const VALIDATION_SYSTEM: &str =
    "You are an analyst evaluating enforcement documents for a healthcare fraud \
     analysis pipeline. You determine whether a document describes a real enforcement \
     action and provide a concise summary.";

type StageResult<T> = Result<T, Box<dyn std::error::Error + Send + Sync>>;

#[derive(Default)]
struct FetchStats {
    fetched: usize,
    skipped: usize,
    failed: usize,
}

enum FetchOutcome {
    Fetched,
    Skipped,
    Failed,
}

/// Extract visible text from HTML content by stripping tags.
pub fn extract_text(html: &str) -> String {
    // Remove script, style, noscript, svg, head elements (case-insensitive)
    let script_re = Regex::new(r"(?is)<script[^>]*>.*?</script>").unwrap();
    let style_re = Regex::new(r"(?is)<style[^>]*>.*?</style>").unwrap();
    let noscript_re = Regex::new(r"(?is)<noscript[^>]*>.*?</noscript>").unwrap();
    let svg_re = Regex::new(r"(?is)<svg[^>]*>.*?</svg>").unwrap();
    let head_re = Regex::new(r"(?is)<head[^>]*>.*?</head>").unwrap();

    let cleaned = script_re.replace_all(html, " ");
    let cleaned = style_re.replace_all(&cleaned, " ");
    let cleaned = noscript_re.replace_all(&cleaned, " ");
    let cleaned = svg_re.replace_all(&cleaned, " ");
    let cleaned = head_re.replace_all(&cleaned, " ");

    let strip_re = Regex::new(r"<[^>]+>").unwrap();
    let text = strip_re.replace_all(&cleaned, " ");
    let lines: Vec<&str> = text
        .lines()
        .map(|l| l.trim())
        .filter(|l| !l.is_empty())
        .collect();
    let result = lines.join("\n");
    let multi_newline = Regex::new(r"\n{3,}").unwrap();
    multi_newline
        .replace_all(&result, "\n\n")
        .trim()
        .to_string()
        .replace('\0', "")
}

/// Detect binary/non-readable content.
pub fn is_binary_content(text: &str) -> bool {
    if text.is_empty() {
        return true;
    }
    let sample = &text[..text.len().min(2000)];
    if sample.trim_start().starts_with("%PDF") {
        return true;
    }
    let non_printable = sample
        .chars()
        .filter(|c| !c.is_alphanumeric() && !c.is_whitespace() && !c.is_ascii_punctuation())
        .count();
    non_printable as f64 / sample.len().max(1) as f64 > 0.15
}

/// Fetch a URL and return the response body as text.
pub async fn fetch_url(url: &str) -> StageResult<String> {
    let client = reqwest::Client::builder()
        .user_agent("SVAP-Pipeline/1.0 (Structural Vulnerability Analysis)")
        .timeout(std::time::Duration::from_secs(30))
        .build()?;
    let resp = client.get(url).send().await?;
    Ok(resp.text().await?)
}

async fn validate_document(
    bedrock: &BedrockClient,
    source_name: &str,
    description: &str,
    url: &str,
    doc_text: &str,
) -> StageResult<(String, bool)> {
    let text_sample = &doc_text[..doc_text.len().min(4000)];
    let prompt = format!(
        "Analyze this enforcement document and provide:\n\
         1. A 2-3 sentence summary of the enforcement action described\n\
         2. Whether this is a valid healthcare fraud enforcement document (true/false)\n\n\
         Document source: {source_name}\n\
         Description: {description}\n\
         URL: {url}\n\n\
         Document text (first portion):\n{text_sample}\n\n\
         Respond in JSON: {{\"summary\": \"...\", \"is_valid\": true/false}}"
    );
    let result = bedrock
        .invoke_json(&prompt, VALIDATION_SYSTEM, Some(0.1), Some(500))
        .await?;
    let summary = result
        .get("summary")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();
    let is_valid = result
        .get("is_valid")
        .and_then(|v| v.as_bool())
        .unwrap_or(false);
    Ok((summary, is_valid))
}

pub async fn run(
    db_client: &Client,
    bedrock: &BedrockClient,
    run_id: &str,
    config: &Config,
) -> StageResult<serde_json::Value> {
    info!("Stage 0: Enforcement Source Fetching");
    db::log_stage_start(db_client, run_id, 0).await?;

    match run_inner(db_client, bedrock, run_id, config).await {
        Ok(result) => {
            db::log_stage_complete(db_client, run_id, 0, Some(&result)).await?;
            Ok(result)
        }
        Err(e) => {
            db::log_stage_failed(db_client, run_id, 0, &e.to_string()).await?;
            Err(e)
        }
    }
}

async fn run_inner(
    db_client: &Client,
    bedrock: &BedrockClient,
    _run_id: &str,
    config: &Config,
) -> StageResult<serde_json::Value> {
    let sources = db::get_enforcement_sources(db_client).await?;
    let ingester = DocumentIngester::new(config);
    let fetch_stats = fetch_missing_sources(db_client, &ingester, &sources).await?;
    let validated = validate_pending_documents(db_client, bedrock).await?;

    let result = json!({
        "documents_fetched": fetch_stats.fetched,
        "documents_skipped": fetch_stats.skipped,
        "documents_failed": fetch_stats.failed,
        "documents_validated": validated,
    });
    info!(
        "Stage 0 complete: {} fetched, {} skipped, {} failed, {} validated.",
        fetch_stats.fetched, fetch_stats.skipped, fetch_stats.failed, validated
    );
    Ok(result)
}

async fn fetch_missing_sources(
    db_client: &Client,
    ingester: &DocumentIngester,
    sources: &[EnforcementSource],
) -> StageResult<FetchStats> {
    let mut stats = FetchStats::default();
    for source in sources {
        match fetch_source_document(db_client, ingester, source).await? {
            FetchOutcome::Fetched => stats.fetched += 1,
            FetchOutcome::Skipped => stats.skipped += 1,
            FetchOutcome::Failed => stats.failed += 1,
        }
    }
    Ok(stats)
}

async fn fetch_source_document(
    db_client: &Client,
    ingester: &DocumentIngester,
    source: &EnforcementSource,
) -> StageResult<FetchOutcome> {
    let Some(url) = source.url.as_deref().filter(|u| !u.is_empty()) else {
        return Ok(FetchOutcome::Skipped);
    };
    if source.has_document {
        return Ok(FetchOutcome::Skipped);
    }

    info!("Fetching: {}", source.name);
    let html = match fetch_url(url).await {
        Ok(html) => html,
        Err(e) => {
            error!("Failed to fetch {}: {}", url, e);
            return Ok(FetchOutcome::Failed);
        }
    };
    let text = extract_text(&html);
    if !is_usable_source_text(&text) {
        return Ok(FetchOutcome::Failed);
    }

    let metadata = json!({
        "url": url,
        "source_name": source.name,
        "source_id": source.source_id,
    });

    let (doc_id, n_chunks) = ingester
        .ingest_text(
            db_client,
            &text,
            &source.source_id,
            "enforcement",
            Some(&metadata),
        )
        .await?;

    let s3_key = format!(
        "enforcement-sources/{}/press_release.html",
        source.source_id
    );
    db::update_enforcement_source_document(db_client, &source.source_id, &s3_key, &doc_id).await?;
    info!("Ingested: {} chars, {} chunks", text.len(), n_chunks);
    Ok(FetchOutcome::Fetched)
}

fn is_usable_source_text(text: &str) -> bool {
    if text.len() < 200 {
        info!("Skipped: text too short ({} chars)", text.len());
        return false;
    }
    if is_binary_content(text) {
        info!("Skipped: binary/non-readable content");
        return false;
    }
    true
}

async fn validate_pending_documents(
    db_client: &Client,
    bedrock: &BedrockClient,
) -> StageResult<usize> {
    let sources = db::get_enforcement_sources(db_client).await?;
    let docs = db::get_all_documents(db_client, Some("enforcement")).await?;
    let mut validated = 0;

    for source in sources.iter().filter(|source| should_validate(source)) {
        if validate_source_document(db_client, bedrock, source, &docs).await? {
            validated += 1;
        }
    }

    Ok(validated)
}

fn should_validate(source: &EnforcementSource) -> bool {
    source.has_document && source.validation_status.as_deref() == Some("pending")
}

async fn validate_source_document(
    db_client: &Client,
    bedrock: &BedrockClient,
    source: &EnforcementSource,
    docs: &[Document],
) -> StageResult<bool> {
    let Some(doc) = docs
        .iter()
        .find(|d| Some(&d.doc_id) == source.doc_id.as_ref())
    else {
        return Ok(false);
    };

    info!("Validating: {}", source.name);
    match validate_document(
        bedrock,
        &source.name,
        source.description.as_deref().unwrap_or("N/A"),
        source.url.as_deref().unwrap_or("uploaded"),
        &doc.full_text,
    )
    .await
    {
        Ok((summary, is_valid)) => {
            let status = if is_valid { "valid" } else { "invalid" };
            db::update_enforcement_source_summary(db_client, &source.source_id, &summary, status)
                .await?;
            Ok(true)
        }
        Err(e) => {
            error!("Validation failed: {}", e);
            db::update_enforcement_source_summary(
                db_client,
                &source.source_id,
                &format!("Validation error: {e}"),
                "error",
            )
            .await?;
            Ok(false)
        }
    }
}
